File: packages/format/src/h265/presentation-order.ts
import {
  isH265IdrNalType,
  isH265RandomAccessNalType
} from "./annex-b.js";
import { requireH265 } from "./failure.js";

export interface H265PictureOrderState {
  initialized: boolean;
  previousTid0PictureOrderCountLsb: number;
  previousTid0PictureOrderCountMsb: number;
}

export interface H265DecodedPictureOrder {
  readonly decodeIndex: number;
  readonly pictureOrderCount: number;
}

export function createH265PictureOrderState(): H265PictureOrderState {
  return {
    initialized: false,
    previousTid0PictureOrderCountLsb: 0,
    previousTid0PictureOrderCountMsb: 0
  };
}

/** Derives PicOrderCntVal with unit-local state (HEVC 8.3.1 subset). */
export function deriveH265PictureOrderCount(
  nalType: number,
  temporalId: number,
  pictureOrderCountLsb: number,
  log2MaxPictureOrderCountLsb: number,
  state: H265PictureOrderState
): number {
  const maximum = 2 ** log2MaxPictureOrderCountLsb;
  requireH265(
    Number.isSafeInteger(pictureOrderCountLsb) &&
      pictureOrderCountLsb >= 0 &&
      pictureOrderCountLsb < maximum,
    "pictureOrderCount",
    "slice picture-order-count LSB is out of range"
  );
  if (isH265IdrNalType(nalType)) {
    state.initialized = true;
    state.previousTid0PictureOrderCountLsb = 0;
    state.previousTid0PictureOrderCountMsb = 0;
    return 0;
  }
  let msb = 0;
  if (state.initialized && !isH265RandomAccessNalType(nalType)) {
    const previousLsb = state.previousTid0PictureOrderCountLsb;
    const previousMsb = state.previousTid0PictureOrderCountMsb;
    if (pictureOrderCountLsb < previousLsb && previousLsb - pictureOrderCountLsb >= maximum / 2) {
      msb = previousMsb + maximum;
    } else if (
      pictureOrderCountLsb > previousLsb &&
      pictureOrderCountLsb - previousLsb > maximum / 2
    ) {
      msb = previousMsb - maximum;
    } else {
      msb = previousMsb;
    }
  }
  const pictureOrderCount = msb + pictureOrderCountLsb;
  // RADL/RASL pictures (types 6..9) do not become prevTid0Pic.
  if (temporalId === 0 && !(nalType >= 6 && nalType <= 9)) {
    state.initialized = true;
    state.previousTid0PictureOrderCountLsb = pictureOrderCountLsb;
    state.previousTid0PictureOrderCountMsb = msb;
  }
  return pictureOrderCount;
}

/** Maps decoder submission order to a contiguous unit-local display order. */
export function deriveH265PresentationOrder(
  pictures: readonly H265DecodedPictureOrder[],
  maximumReorderPictures: number,
  path: string
): readonly number[] {
  requireH265(pictures.length > 0, path, "unit contains no decoded pictures");
  const sorted = [...pictures].sort(
    (left, right) => left.pictureOrderCount - right.pictureOrderCount
  );
  const first = sorted[0]?.pictureOrderCount;
  requireH265(first !== undefined, path, "unit contains no presentation pictures");
  const decodeToPresentation = new Array<number>(pictures.length);
  for (let presentationIndex = 0; presentationIndex < sorted.length; presentationIndex += 1) {
    const picture = sorted[presentationIndex];
    requireH265(picture !== undefined, path, "presentation picture is missing");
    requireH265(
      picture.pictureOrderCount === first + presentationIndex,
      path,
      "unit picture-order counts must be unique and contiguous"
    );
    requireH265(
      Number.isSafeInteger(picture.decodeIndex) &&
        picture.decodeIndex >= 0 &&
        picture.decodeIndex < pictures.length &&
        decodeToPresentation[picture.decodeIndex] === undefined,
      path,
      "unit decode index is duplicated or out of range"
    );
    decodeToPresentation[picture.decodeIndex] = presentationIndex;
  }
  let requiredReorder = 0;
  for (let decodeIndex = 0; decodeIndex < decodeToPresentation.length; decodeIndex += 1) {
    const presentationIndex = decodeToPresentation[decodeIndex];
    requireH265(presentationIndex !== undefined, path, "decode order has a gap");
    requiredReorder = Math.max(requiredReorder, decodeIndex - presentationIndex);
  }
  requireH265(
    requiredReorder <= maximumReorderPictures,
    path,
    "derived presentation reordering exceeds the SPS declaration"
  );
  return Object.freeze(decodeToPresentation);
}

File: flutter/packages/aval_format/lib/src/h265/presentation_order.dart
/// HEVC picture-order-count derivation and unit-local presentation ordering.
///
/// Dart port of `packages/format/src/h265/presentation-order.ts`.
library;

import 'annex_b.dart' show isH265IdrNalType, isH265RandomAccessNalType;
import 'failure.dart';

/// Mutable per-unit POC derivation state.
///
/// Port of `H265PictureOrderState` (`src/h265/presentation-order.ts:7`).
class H265PictureOrderState {
  H265PictureOrderState({
    required this.initialized,
    required this.previousTid0PictureOrderCountLsb,
    required this.previousTid0PictureOrderCountMsb,
  });

  bool initialized;
  int previousTid0PictureOrderCountLsb;
  int previousTid0PictureOrderCountMsb;
}

/// Port of `H265DecodedPictureOrder` (`src/h265/presentation-order.ts:13`).
class H265DecodedPictureOrder {
  const H265DecodedPictureOrder({
    required this.decodeIndex,
    required this.pictureOrderCount,
  });

  final int decodeIndex;
  final int pictureOrderCount;
}

/// Port of `createH265PictureOrderState`
/// (`src/h265/presentation-order.ts:18`).
H265PictureOrderState createH265PictureOrderState() {
  return H265PictureOrderState(
    initialized: false,
    previousTid0PictureOrderCountLsb: 0,
    previousTid0PictureOrderCountMsb: 0,
  );
}

/// Derives PicOrderCntVal with unit-local state (HEVC 8.3.1 subset).
///
/// Port of `deriveH265PictureOrderCount`
/// (`src/h265/presentation-order.ts:27`).
int deriveH265PictureOrderCount(
  int nalType,
  int temporalId,
  int pictureOrderCountLsb,
  int log2MaxPictureOrderCountLsb,
  H265PictureOrderState state,
) {
  final maximum = 1 << log2MaxPictureOrderCountLsb;
  requireH265(
    pictureOrderCountLsb >= 0 && pictureOrderCountLsb < maximum,
    'pictureOrderCount',
    'slice picture-order-count LSB is out of range',
  );
  if (isH265IdrNalType(nalType)) {
    state.initialized = true;
    state.previousTid0PictureOrderCountLsb = 0;
    state.previousTid0PictureOrderCountMsb = 0;
    return 0;
  }
  var msb = 0;
  if (state.initialized && !isH265RandomAccessNalType(nalType)) {
    final previousLsb = state.previousTid0PictureOrderCountLsb;
    final previousMsb = state.previousTid0PictureOrderCountMsb;
    if (pictureOrderCountLsb < previousLsb &&
        previousLsb - pictureOrderCountLsb >= maximum / 2) {
      msb = previousMsb + maximum;
    } else if (pictureOrderCountLsb > previousLsb &&
        pictureOrderCountLsb - previousLsb > maximum / 2) {
      msb = previousMsb - maximum;
    } else {
      msb = previousMsb;
    }
  }
  final pictureOrderCount = msb + pictureOrderCountLsb;
  // RADL/RASL pictures (types 6..9) do not become prevTid0Pic.
  if (temporalId == 0 && !(nalType >= 6 && nalType <= 9)) {
    state.initialized = true;
    state.previousTid0PictureOrderCountLsb = pictureOrderCountLsb;
    state.previousTid0PictureOrderCountMsb = msb;
  }
  return pictureOrderCount;
}

/// Maps decoder submission order to a contiguous unit-local display order.
///
/// Port of `deriveH265PresentationOrder`
/// (`src/h265/presentation-order.ts:74`).
List<int> deriveH265PresentationOrder(
  List<H265DecodedPictureOrder> pictures,
  int maximumReorderPictures,
  String path,
) {
  requireH265(pictures.isNotEmpty, path, 'unit contains no decoded pictures');
  final sorted = List<H265DecodedPictureOrder>.from(pictures)
    ..sort((left, right) => left.pictureOrderCount - right.pictureOrderCount);
  final first = sorted[0].pictureOrderCount;
  final decodeToPresentation = List<int?>.filled(pictures.length, null);
  for (var presentationIndex = 0;
      presentationIndex < sorted.length;
      presentationIndex += 1) {
    final picture = sorted[presentationIndex];
    requireH265(
      picture.pictureOrderCount == first + presentationIndex,
      path,
      'unit picture-order counts must be unique and contiguous',
    );
    requireH265(
      picture.decodeIndex >= 0 &&
          picture.decodeIndex < pictures.length &&
          decodeToPresentation[picture.decodeIndex] == null,
      path,
      'unit decode index is duplicated or out of range',
    );
    decodeToPresentation[picture.decodeIndex] = presentationIndex;
  }
  var requiredReorder = 0;
  for (var decodeIndex = 0;
      decodeIndex < decodeToPresentation.length;
      decodeIndex += 1) {
    final presentationIndex = decodeToPresentation[decodeIndex];
    requireH265(presentationIndex != null, path, 'decode order has a gap');
    final delta = decodeIndex - presentationIndex!;
    requiredReorder = requiredReorder > delta ? requiredReorder : delta;
  }
  requireH265(
    requiredReorder <= maximumReorderPictures,
    path,
    'derived presentation reordering exceeds the SPS declaration',
  );
  return List.unmodifiable(decodeToPresentation.cast<int>());
}
